spec/operators/toSet-spec.ts

```diff
@@ -0,0 +1,112 @@
+    const e1 =  cold('-');
+    const e1subs =   '^';
+    const expected = '-';
+
+    expectObservable(e1.toSet()).toBe(expected);
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should be never when source is empty', () => {
+    const e1 =  cold('|');
+    const e1subs =   '(^!)';
+    const expected = '(w|)';
+
+    expectObservable(e1.toSet()).toBe(expected, { w: new Set() });
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should be never when source doesn\'t complete', () => {
+    const e1 = hot('--x--^--y--');
+    const e1subs =      '^     ';
+    const expected =    '------';
+
+    expectObservable(e1.toSet()).toBe(expected);
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should reduce observable without values into an set of size zero', () => {
+    const e1 = hot('-x-^---|');
+    const e1subs =    '^   !';
+    const expected =  '----(w|)';
+
+    expectObservable(e1.toSet()).toBe(expected, { w: new Set() });
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should reduce the a single value of an observable into an set', () => {
+    const e1 = hot('-x-^--y--|');
+    const e1subs =    '^     !';
+    const expected =  '------(w|)';
+
+    expectObservable(e1.toSet()).toBe(expected, { w: new Set(['y']) });
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should allow unsubscribing explicitly and early', () => {
+    const e1 =   hot('--a--b----c-----d----e---|');
+    const unsub =    '        !                 ';
+    const e1subs =   '^       !                 ';
+    const expected = '---------                 ';
+
+    expectObservable(e1.toSet(), unsub).toBe(expected);
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should not break unsubscription chains when result is unsubscribed explicitly', () => {
+    const e1 =   hot('--a--b----c-----d----e---|');
+    const e1subs =   '^       !                 ';
+    const expected = '---------                 ';
+    const unsub =    '        !                 ';
+
+    const result = e1
+      .mergeMap((x: string) => Observable.of(x))
+      .toSet()
+      .mergeMap((x: Array<string>) => Observable.of(x));
+
+    expectObservable(result, unsub).toBe(expected);
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should work with error', () => {
+    const e1 = hot('-x-^--y--z--#', { x: 1, y: 2, z: 3 }, 'too bad');
+    const e1subs =    '^        !';
+    const expected =  '---------#';
+
+    expectObservable(e1.toSet()).toBe(expected, null, 'too bad');
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  it('should work with throw', () => {
+    const e1 =  cold('#');
+    const e1subs =   '(^!)';
+    const expected = '#';
+
+    expectObservable(e1.toSet()).toBe(expected);
+    expectSubscriptions(e1.subscriptions).toBe(e1subs);
+  });
+
+  type(() => {
```

Comment:
this isn't necessary in general unless tricky type definition verification is required.